src/Compiler/AbstractIL/ilmorph.fs

```diff
@@ -256,8 +256,21 @@ let morphILMethodDefs f (m: ILMethodDefs) = mkILMethods (List.map f (m.AsList())
 let morphILFieldDefs f (fdefs: ILFieldDefs) =
     mkILFields (List.map f (fdefs.AsList()))
 
-let morphILTypeDefs f (tdefs: ILTypeDefs) =
-    mkILTypeDefsFromArray (Array.map f (tdefs.AsArray()))
+let morphILTypeDefs isInKnownSet f (tdefs: ILTypeDefs) =
+    let filtered (tdefs: ILTypeDef array) =
+        let seen = new HashSet<string>()
+
+        [|
+            for td in tdefs do
+                match isInKnownSet td.Name with
+                | true when not (seen.Contains td.Name) ->
+                    seen.Add td.Name |> ignore
+                    yield td
+                | true -> ()
+                | false -> yield td
+        |]
```

Comment:
Do we need to preserve order here?

If not we could do:
```fs
let filtered (tdefs: ILTypeDef array) =
    let inKnownSet, rest = Array.partition (fun td -> isInKnownSet td.Name) tdefs
    inKnownSet
    |> Array.distinctBy (fun td -> td.Name)
    |> Array.concat rest
```
I would find that more clear, but if the order matters OR you disagree about readability, then feel free to ignore.

Reply:
Typedef order is probably not important, except for all of the baseline files which may have their types reordered.  I'm not sure although, I expect if I could come up with an improved filter.